src/core/wee-eval.c

```diff
@@ -1070,6 +1070,28 @@ eval_replace_vars_cb (void *data, const char *text)
         ptr_value = hashtable_get (secure_hashtable_data, text + 9);
         return strdup ((ptr_value) ? ptr_value : "");
     }
+    if (strncmp (text, "sec.cmd.", 8) == 0)
+    {
+        const char *program;
+        char *value, *status;
+        FILE *in;
+
+        status = NULL;
+        if (!(program = hashtable_get (secure_hashtable_cmds, text + 8)))
+        {
+            return strdup("");
+        }
+        if (!(in = popen (program, "r")))
```

Comment:
Instead of using `popen` directly, WeeChat has `hook_process` and `hook_process_hashtable` for executing external processes. They take care of all the intricacies of managing child processes in an asynchronous way and then the child process is executed in a way that shouldn't mess with WeeChat's own terminal.

That does, however, lead to the core issue: eval needs to return the output from executed process immediately (in a blocking way) instead of asynchronously (which is the WeeChat's way of running external processes).
Currently it seems that the process must output something fast enough but that's a massive data race, no? If the process takes a moment to initialize, sync or decrypt some passwords container, it'd already be too late. On the other hand, WeeChat can't just wait for output in the middle of evaluating an expression because it'd block WeeChat completely.

I think the main reason why this has stayed unsolved so far is that there doesn't seem to be an easy solution to these issues.

Reply:
Ahh I see.

I think we have the following options to deal with the problem.

1. Run the process async. The expression will just return "" until the value is ready. The value could be cached to make sure that this does not occur too often. We could even read the values at startup and cache them for as long weechat runs.

2. Block weechat entirely. This would either happen at startup time (And the programm output is just cached) or when the user requests the value.

Im not really sure if blocking is such a big problem. Yes weechat will hang. But this happens in response to a user initiated action. Requesting the password. I don't think the User is really suprised when weechat hangs until the process returns.

Reply:
1. Just returning `""` initially would be a problem because a password may be required to be available immediately on WeeChat startup, e.g. autoconnecting IRC server or relay server.
2. Caching values is also problematic because then we have to do cache invalidation somehow. Users may change the passwords in their password manager.
3. By caching passwords we somewhat duplicate the function of the password manager and decrease security by having copies of passwords also in WeeChat memory. Proper password managers take security very seriously and use special memory protection techniques to prevent leaks through RAM.
4. Moreover, we can't read values at startup ahead of time (even if we ignore point 1.) because we don't know which eval expressions will be executed later in time. And the queried values may also depend on anything else dynamically, so it's an uphill battle to try to do that.
5. Also, if you're suggesting reading values at startup and caching them, you might as well write a script/plugin that (given a list of password names) queries them (asynchronously) from the password manager and sets them in `/secure`. It still has the issue from point 1.
6. Blocking WeeChat is a big problem, user input or not. If WeeChat is blocked waiting for an external process, IRC connections etc can time out (WeeChat isn't responding to `PING`s at that time). Also all `hook_timer`s would majorly mess up. Basically, the entirety of WeeChat has been designed in an asynchronous manner and changing that means changing everything.